File: servico-musica-java/java.grpc/src/main/java/servico/java/grpc/controller/PlaylistController.java
package servico.java.grpc.controller;

import io.grpc.stub.StreamObserver;
import jakarta.transaction.Transactional;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.grpc.server.service.GrpcService;

import servico.java.grpc.entity.Playlist;
import servico.java.grpc.entity.Usuario;
import servico.java.grpc.proto.PlaylistListResponse;
import servico.java.grpc.proto.PlaylistResponse;
import servico.java.grpc.proto.PlaylistServiceGrpc;
import servico.java.grpc.proto.PlaylistPorUsuarioRequest;
import servico.java.grpc.proto.PlaylistPorMusicaRequest;
import servico.java.grpc.repository.PlaylistRepository;
import servico.java.grpc.repository.UsuarioRepository;

import java.util.List;
import java.util.Optional;

@GrpcService
public class PlaylistController extends PlaylistServiceGrpc.PlaylistServiceImplBase {

    @Autowired
    private PlaylistRepository playlistRepository;

    @Autowired
    private UsuarioRepository usuarioRepository;

    @Transactional
    @Override
    public void playlistsPorUsuario(PlaylistPorUsuarioRequest request, StreamObserver<PlaylistListResponse> responseObserver) {

        PlaylistListResponse.Builder response = PlaylistListResponse.newBuilder();

        Optional<Usuario> usuario = usuarioRepository.findById(request.getIdUsuario());

        if (usuario.isPresent()) {
            for (Playlist playlist : usuario.get().getPlaylistsUsuario()) {

                response.addPlaylists(
                        PlaylistResponse.newBuilder()
                                .setId(playlist.getId())
                                .setNome(playlist.getNome())
                                .build()
                );
            }
        }

        responseObserver.onNext(response.build());
        responseObserver.onCompleted();
    }

    @Override
    public void playlistsPorMusica(PlaylistPorMusicaRequest request, StreamObserver<PlaylistListResponse> responseObserver) {

        List<Playlist> playlists = playlistRepository.findByMusicasId(request.getIdMusica());

        PlaylistListResponse.Builder response = PlaylistListResponse.newBuilder();

        for (Playlist playlist : playlists) {
            response.addPlaylists(
                    PlaylistResponse.newBuilder()
                            .setId(playlist.getId())
                            .setNome(playlist.getNome())
                            .build()
            );
        }

        responseObserver.onNext(response.build());
        responseObserver.onCompleted();
    }
}

File: servico-musica-kotlin/kotlin.grpc/src/main/kotlin/servico/kotlin/grpc/controller/PlaylistController.kt
package servico.kotlin.grpc.controller

import io.grpc.stub.StreamObserver
import jakarta.transaction.Transactional
import org.springframework.grpc.server.service.GrpcService
import servico.kotlin.grpc.entity.Playlist
import servico.kotlin.grpc.entity.Usuario
import servico.kotlin.grpc.proto.*
import servico.kotlin.grpc.repository.PlaylistRepository
import servico.kotlin.grpc.repository.UsuarioRepository
import java.util.*

@GrpcService
class PlaylistController(
    private val playlistRepository: PlaylistRepository,
    private val usuarioRepository: UsuarioRepository
) : PlaylistServiceGrpc.PlaylistServiceImplBase() {

    @Transactional
    override fun playlistsPorUsuario(
        request: PlaylistPorUsuarioRequest,
        responseObserver: StreamObserver<PlaylistListResponse>
    ) {

        val response = PlaylistListResponse.newBuilder()

        val usuario: Optional<Usuario> =
            usuarioRepository.findById(request.idUsuario)

        if (usuario.isPresent) {
            usuario.get().playlistsUsuario.forEach { playlist ->
                response.addPlaylists(
                    PlaylistResponse.newBuilder()
                        .setId(playlist.id)
                        .setNome(playlist.nome)
                        .build()
                )
            }
        }

        responseObserver.onNext(response.build())
        responseObserver.onCompleted()
    }

    override fun playlistsPorMusica(
        request: PlaylistPorMusicaRequest,
        responseObserver: StreamObserver<PlaylistListResponse>
    ) {

        val playlists: List<Playlist> =
            playlistRepository.findByMusicasId(request.idMusica)

        val response = PlaylistListResponse.newBuilder()

        playlists.forEach { playlist ->
            response.addPlaylists(
                PlaylistResponse.newBuilder()
                    .setId(playlist.id)
                    .setNome(playlist.nome)
                    .build()
            )
        }

        responseObserver.onNext(response.build())
        responseObserver.onCompleted()
    }
}